### ISSUE TITLE
Cannot have more than one submit button in an app working correctly

### ISSUE THREAD

**Issue by daattali** [rstudio/shiny#1426]
I've never really used submit buttons but I was trying to use them tonight and noticed that if you have more than one submit button in an app, they ignore their enclosing `<form>` and they act as if the entire app is one form. Submit buttons should be isolated within each form, but it looks like activating any submit in an app will cause all inputs in the app to fire.

I also noticed that even if I create a submit button with pure HTML I still run into the issue, so the problem is probably something weird that shiny is doing.

Example:

```
library(shiny)

ui <- fluidPage(
  tags$form(
    textInput("text1", "text1"),
    tags$button("submit1", type="submit", class="btn btn-default")
  ),
  tags$form(
    textInput("text2", "text2"),
    tags$button("submit2", type="submit", class="btn btn-default")
  )
)

server <- function(input, output, session) {
  observe({
    input$text1
    cat("submit 1")
  })
  observe({
    input$text2
    cat("submit 2")
  })
}

shinyApp(ui = ui, server = server)
```

A related question is: can we deprecate submit buttons since they're so awkward (they don't even take an id!), that would be the best :)

**Comment by bborgesr** [comment]
Hi Dean,
Yes, I think submit buttons are Shiny's black sheep child. They were never very good, they have lots of small issues and action buttons can do everything (and do it better) that you might want to do with submit buttons. Here a similar issue that Winston filed: #303. I think our current policy is that we discourage their use without actually deprecating them (mostly because of backwards compatibility, I think) -- see [Joe's comment here](https://github.com/rstudio/shiny/issues/95#issuecomment-238316087). But the bottom line is the same, I think: we're keeping it for compatibility reasons, and are unlikely to spend much, if any, time on it...

For your simple app, here's a similar app using action buttons instead (for more complicated apps, you might need to get into `isolate`, as I'm sure you know):

```r
library(shiny)

ui <- fluidPage(
  tags$form(
    textInput("text1", "text1"),
    tags$button("submit1", id = "submit1", type="button",
                class="btn btn-default action-button")
  ),
  tags$form(
    textInput("text2", "text2"),
    tags$button("submit2", id = "submit2", type="button",
                class="btn btn-default action-button")
  )
)

server <- function(input, output, session) {
  observeEvent(input$submit1, {
    cat(input$text1)
  })
  observeEvent(input$submit2, {
    cat(input$text2)
  })
}

shinyApp(ui = ui, server = server)
```

**Comment by daattali** [comment]
Thanks @bborgesr 

This bug was more of an FYI, I don't actually need to have multiple submits.

@jcheng5 said you can't deprecate it (because unfortunately it is used in a lot of apps...) but it's discouraged. I don't think the average shiny user realizes it's a weird button to use - perhaps the documentation for submitButton could mention that it's not recommended

**Comment by bborgesr** [comment]
What do you say @jcheng5? Should we update the documentation like @daattali is suggesting (and while we're at it, also address #1425)? We could also copy or link to the app above (as is or slightly modified) in order to show how action buttons can perform the role of submit buttons (and highlight that this may even be conceptually easier for app authors because they would only have to worry about one kind of button)...

**Comment by bborgesr** [comment]
Note to self: fix together with #1355
